Eric install setup

darth-cheney opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
Minor adjustments to `setup.py` build process and creation of initial install / develop instructions
## Motivation and Context
### Setup File
When using the plain `setup.py` file for building and installation, there was an issue whereby `numpy` needed to be installed before the build (due to nativepython needing the header files) but also being a subsequent package dependency. We accomplish this by setting the build extension to its own context class.
  
### Installation Instructions
As it stands, there is no document that describes the best or known method(s) for: building, installing, building for development, or installing for development. We have included a **preliminary** version of this document.

## Approach
### Setup file
For the `setup.py` adjustments, it was simply a matter of creating a new build extension subclass and having that be the command class during the setup process. This ensures that in fresh environments the `numpy` headers will be available before the build process.
  
## Installation Instructions
We used a combination of asking questions and running processes ourselves to come up with this current preliminary version of the instructions. We have done our best to test these procedures on OSX and Ubuntu so far.

## How Has This Been Tested?
After any changes, we re-ran the included tests (since we did not write any new meaningful code).

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

braxtonmckee commented on INSTALLATION.md:
````diff
@@ -0,0 +1,48 @@
+  Note that for development you will also install the debug interpreter.
+  On Ubuntu this is not so straight forward. You will first need to add the following PPA:
+  ```
+  sudo add-apt-repository ppa:jonathonf/python-3.6
````
for which ubuntu? I'm on 18.04 and it worked out of the box for me

darth-cheney replied:
True I was testing on system that run 16, I will make a note of that

atzannes commented on INSTALLATION.md:
```diff
@@ -0,0 +1,48 @@
+### Pipenv Method ###
+1. (Optional) Create a new virtualenv with Python 3.6 (`virtualenv --python=<path-to-py3> venv`) and source it. If you choose to use Pipenv alone, it will create an appropriate virtualenv for you.
+2. Run `pipenv install`
+3. Append the root directory of this repository to your `PYTHONPATH`
```
I think that this 3rd step is not necessary because `nativepython` is installed in development mode (see the line in `Pipfile` that reads `nativepython = {editable = true,path = "."}`.

darth-cheney replied:
Yes, true, that makes sense. Will change.

atzannes commented on INSTALLATION.md:
```diff
@@ -0,0 +1,48 @@
+  
+### Pipenv Method ###
+1. (Optional) Create a new virtualenv with Python 3.6 (`virtualenv --python=<path-to-py3> venv`) and source it. If you choose to use Pipenv alone, it will create an appropriate virtualenv for you.
+2. Run `pipenv install`
```
I think that this should be `pipenv install --dev --deploy`, like in the `Makefile`
- `--dev` also installs the required packages for running the tests
- `--deploy` will prevent `pipenv` from updating its lock file, so it will force it to install the versions of the dependencies specified therein or fail.

darth-cheney replied:
You're right — will update

atzannes commented on INSTALLATION.md:
```diff
@@ -0,0 +1,48 @@
+4. Append the root of this repository to your `PYTHONPATH`
+  
+### Pipenv Method ###
+1. (Optional) Create a new virtualenv with Python 3.6 (`virtualenv --python=<path-to-py3> venv`) and source it. If you choose to use Pipenv alone, it will create an appropriate virtualenv for you.
```
Since the `Makefile` implements this workflow, should we document how to use that as well or instead of these steps? Right now, `make install` should perform steps 1 and 2 of these instructions. Feel free to change the default python parameter of the `Makefile` from `python3` to `python3.6` on line 5

darth-cheney replied:
Yes, I will add a section for the Makefile methods and note a caveat about the python versions

atzannes commented on setup.py:
```diff
@@ -33,52 +58,13 @@ def is_numpy_installed():
-        'lz4',
+    cmdclass={'build_ext': NumpyBuildExtension},
+    ext_modules=ext_modules,
+    setup_requires=[
```
nice!

atzannes commented on setup.py:
```diff
@@ -33,52 +58,13 @@ def is_numpy_installed():
-        'redis',
-        'requests',
-        'websockets',
+        'flask'
```
I know why setup requires numpy, but why does it require flask?

darth-cheney replied:
flask and friends are required in order to properly run the tests (object database and web tools use it). We can remove them from this setup file, I think, but we'd have to keep them in the requirements/pipfile

atzannes commented on requirements.txt:
```diff
@@ -0,0 +1,21 @@
+numpy
```
numpy and flask also appear later in this list.

atzannes commented:
It looks like you may need to manually rebase on top of `dev`. I also see that the lint tests have failed. I have already fixed some of them in dev, but to ignore the errors under `.egg` I would add that directory to the list of ignored ones in `tox.ini`

darth-cheney commented:
Hey @braxtonmckee — I pushed some changes to the `eric_install_setup` branch that should resolve all outstanding issues here (I have also ran installation and tests on a fresh AWS instance with no errors).
  
It seems you merged the `dev` branch once already, probably on Friday before we got to talking about more changes. Would you like me to run the manual merge, or rebase, or anything? I don't just want to force push changes to `dev` since you seem to have a lot of activity there.
  
@atzannes I think I've addressed all of your issues here. Maybe let's get this one merged in and if you see anything else you want to highlight we can move it over to issues and deal with it as we go, sound good?

braxtonmckee commented:
Yes, I pushed some code to dev. Can you rebase on top? Generally, trying to
keep the history linear unless the rebase is so bad that we can't.

On Mon, Mar 18, 2019 at 11:38 AM Eric Gade <notifications@github.com> wrote:

> Hey @braxtonmckee <https://github.com/braxtonmckee> — I pushed some
> changes to the eric_install_setup branch that should resolve all
> outstanding issues here (I have also ran installation and tests on a fresh
> AWS instance with no errors).
>
> It seems you merged the dev branch once already, probably on Friday
> before we got to talking about more changes. Would you like me to run the
> manual merge, or rebase, or anything? I don't just want to force push
> changes to dev since you seem to have a lot of activity there.
>
> @atzannes <https://github.com/atzannes> I think I've addressed all of
> your issues here. Maybe let's get this one merged in and if you see
> anything else you want to highlight we can move it over to issues and deal
> with it as we go, sound good?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/braxtonmckee/nativepython/pull/57#issuecomment-473964812>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHzjhGcmbtIW38AsNib6B5v1YKhEC7bcks5vX7LpgaJpZM4b25e->
> .
>

darth-cheney commented:
@braxtonmckee What procedure to you want to use for this and future PRs? Do you want to have the final say and actually conduct the merge, or do you prefer that once you approve it the incoming branch owner should handle it?
